src/ts_catalog/metadata.c

```diff
@@ -156,8 +156,13 @@ ts_metadata_insert(const char *metadata_key, Datum metadata_value, Oid type,
 	}
 
 	/* We have to copy the key here because heap_form_tuple will copy NAMEDATALEN
-	 * into the tuple instead of checking length. */
-	strlcpy(key_data, metadata_key, NAMEDATALEN);
```

Comment:
could you please remove this part as `strlcpy` should take care of the trailing zero - I've checked with valgrind and this doesn't cause/fix any issue

Reply:
I'd like to defend this change.

Firstly, the first patch and the second patch solve different problems, that are only grouped in one issue because they are were found by valgrind together.

The thing with strlcpy is that it only copies amount of bytes up to length of source string and then terminates the string with zero. 
In this case, key_data is a char array the size of NAMEDATALEN (which is 64). And metadata_key that is being copied there isn't always NAMEDATALEN long. And key_data is allocated at the top of the stack, and can contain junk. Or whatever was at the top of the stack. So it's better to fill it it with zeroes and only then copy something there.

For me, valgrind still shows first error stated in #5311, and pageinspect still shows junk. And with patch (initial or the ones linked below) this junk is gone and there are only zeroes.

Reproduction is simple - start postgres under valgrind and execute 'CREATE EXTENSION timescaledb;' with psql.

I can provide at least two alternative patches that will first fill the string with zeroes: see attachements

[metadata_memory_fix_memcpy.patch](https://github.com/timescale/timescaledb/files/10731963/metadata_memory_fix_memcpy.patch)
[metadata_memory_fix_namedata.patch](https://github.com/timescale/timescaledb/files/10731962/metadata_memory_fix_namedata.patch)

If you'll like any of above patches, I'll replace current code in commit with that patch.

If you're still not convinced and can't reproduce it, I'll remove changes in metadata.c.

Reply:
You actually have a very good point: the value for the `Name` field is a fixed-size char array (effectively, at least), which means that there will be random data added to the tuple.

Note that this has nothing to do with whether the field is dynamically allocated or not, it is just a side-effect of copying `NAMEDATALEN` bytes from the source.

However, your patch does not solve the problem in that case, since all you do is replicating the behavior of `strlcpy` with this change.

I agree that a working change would be what you suggest: initialize the `key_data` variable with zeroes (it is an auto variable, so it is not automatically [zero initialized](https://en.cppreference.com/w/cpp/language/zero_initialization)) and then use `strlcpy`. You need to use `strlcpy` since you need to guarantee that the string is null-terminated even if `metadata_key` is too long for the buffer.

Reply:
> Please let me chime in. As [strncpy(3)](https://man7.org/linux/man-pages/man3/strncpy.3.html) says:
>>    If the length of src is less than n, strncpy() writes additional null bytes to dest to ensure that a total of n bytes are written.

Thank you, I missed that comment. In that case, moving away from `strlcpy` is an improvement, but I agree that using `namestrcpy is the better alternative.